airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumMessageProducer.java

```diff
@@ -0,0 +1,138 @@
+import org.slf4j.LoggerFactory;
+
+public class DebeziumMessageProducer<T> implements SourceStateMessageProducer<ChangeEventWithMetadata> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DebeziumMessageProducer.class);
+
+  private final CdcStateHandler cdcStateHandler;
+
+  /**
+   * `checkpointOffsetToSend` is used as temporal storage for the offset that we want to send as
+   * message. As Debezium is reading records faster that we process them, if we try to send
+   * `offsetManger.read()` offset, it is possible that the state is behind the record we are currently
+   * propagating. To avoid that, we store the offset as soon as we reach the checkpoint threshold
+   * (time or records) and we wait to send it until we are sure that the record we are processing is
+   * behind the offset to be sent.
+   */
+  private final HashMap<String, String> checkpointOffsetToSend = new HashMap<>();
+
+  /**
+   * `previousCheckpointOffset` is used to make sure we don't send duplicated states with the same
+   * offset. Is it possible that the offset Debezium report doesn't move for a period of time, and if
+   * we just rely on the `offsetManger.read()`, there is a chance to sent duplicate states, generating
+   * an unneeded usage of networking and processing.
+   */
+  private final HashMap<String, String> initialOffset, previousCheckpointOffset;
+  private final AirbyteFileOffsetBackingStore offsetManager;
+  private final CdcTargetPosition<T> targetPosition;
+  private final Optional<AirbyteSchemaHistoryStorage> schemaHistoryManager;
+
+  private boolean shouldEmitStateMessage = false;
+
+  private final DebeziumEventConverter eventConverter;
+
+  public DebeziumMessageProducer(
+                                 final CdcStateHandler cdcStateHandler,
+                                 final CdcTargetPosition targetPosition,
+                                 final DebeziumEventConverter eventConverter,
+                                 final AirbyteFileOffsetBackingStore offsetManager,
+                                 final Optional<AirbyteSchemaHistoryStorage> schemaHistoryManager) {
+    this.cdcStateHandler = cdcStateHandler;
+    this.targetPosition = targetPosition;
+    this.eventConverter = eventConverter;
+    this.offsetManager = offsetManager;
+    this.schemaHistoryManager = schemaHistoryManager;
+    this.previousCheckpointOffset = (HashMap<String, String>) offsetManager.read();
+    this.initialOffset = new HashMap<>(this.previousCheckpointOffset);
+    resetCheckpointValues();
+  }
+
+  @Override
+  public AirbyteStateMessage generateStateMessageAtCheckpoint(ConfiguredAirbyteStream stream) {
+    LOGGER.info("Sending CDC checkpoint state message.");
+    final AirbyteStateMessage stateMessage = createStateMessage(checkpointOffsetToSend);
+    previousCheckpointOffset.clear();
+    previousCheckpointOffset.putAll(checkpointOffsetToSend);
+    shouldEmitStateMessage = false;
+    return stateMessage;
```

Comment:
should we call `resetCheckpointValues()` i.e. call `checkpointOffsetToSend.clear();` before returning a state message? 

Also we can probably replace  `resetCheckpointValues()` with checkpointOffsetToSend.clear()`

Reply:
that's a great catch, thanks! that actually solved test issue I've had since I turned up the mssql. Surprisingly tests on postgres did not capture that